src/Umbraco.Web/PublishedCache/IContentRouter.cs

```diff
@@ -0,0 +1,59 @@
+        /// <param name="route">The route</param>
+        /// <param name="hideTopLevelNode">A value forcing the HideTopLevelNode setting.</param>
+        /// <returns>The content, or null.</returns>
+        /// <remarks>
+        /// <para>A valid route is either a simple path eg <c>/foo/bar/nil</c> or a root node id and a path, eg <c>123/foo/bar/nil</c>.</para>
+        /// <para>If <param name="hideTopLevelNode" /> is <c>null</c> then the settings value is used.</para>
+        /// <para>The value of <paramref name="preview"/> overrides defaults.</para>
+        /// </remarks>
+        ContentRoutingResult GetIdByRoute(bool preview, string route, bool? hideTopLevelNode, string culture);
+        /// <summary>
+        /// Gets content id identified by a route.
+        /// </summary>
+        /// <param name="route">The route</param>
+        /// <param name="hideTopLevelNode">A value forcing the HideTopLevelNode setting.</param>
+        /// <returns>The content, or null.</returns>
+        /// <remarks>
+        /// <para>A valid route is either a simple path eg <c>/foo/bar/nil</c> or a root node id and a path, eg <c>123/foo/bar/nil</c>.</para>
+        /// <para>If <param name="hideTopLevelNode" /> is <c>null</c> then the settings value is used.</para>
+        /// <para>Considers published or unpublished content depending on defaults.</para>
+        /// </remarks>
+        ContentRoutingResult GetIdByRoute(string route, bool? hideTopLevelNode, string culture);
+        /// <summary>
+        /// Gets the route for a content identified by its unique identifier.
+        /// </summary>
+        /// <param name="preview">A value indicating whether to consider unpublished content.</param>
+        /// <param name="contentId">The content unique identifier.</param>
+        /// <returns>The route.</returns>
+        /// <remarks>The value of <paramref name="preview"/> overrides defaults.</remarks>
+        string GetRouteById(bool preview, int contentId, string culture = null);
+        // <summary>
+        /// Gets the route for a content identified by its unique identifier.
+        /// </summary>
+        /// <param name="contentId">The content unique identifier.</param>
+        /// <returns>The route.</returns>
+        /// <remarks>Considers published or unpublished content depending on defaults.</remarks>
+        string GetRouteById(int contentId, string culture = null);
+
+        /// <summary>
+        /// Gets the route for a content identified by its alias.
+        /// </summary>
+        /// <param name="preview">In Preview</param>
+        /// <param name="rootNodeId">Domain content id</param>
+        /// <param name="culture">Request culture</param>
+        /// <param name="alias">Url Alias</param>
+        /// <returns>Routing Outcome</returns>
+        ContentRoutingResult GetIdByAlias(bool preview, int rootNodeId, string culture, string alias);
```

Comment:
(Brain dump...)

This abstraction will be able to replace the logic for the built in IContentFinders which is nice but any custom IContentFinders will still directly access the cache. As an example, Articulate has a custom one: https://github.com/Shazwazza/Articulate/blob/master/src/Articulate/Routing/DateFormattedPostContentFinder.cs ... that's a pretty custom one that needs to query the content cache for property values. I guess IContentFinders can be replaced so there is still an opportunity for devs to replace custom ones too. In the future, or as an increment to this work, I think it would be good to have routes persisted to a DB table and that persisting would occur when content is saved where we could ask all IUrlProvider for the content's URLs and save those. In that case IContentFinders won't really execute, only as a last resort if no persisted URLs are matched. 

Some notes on existing abstractions:

* IPublishedRouter - which is a large service and too big to re-implement nicely but potentially the logic within it's `bool TryRouteRequest(PublishedRequest request)` is what needs to be examined and could be abstracted. This does a few things:
  * Looks up domains - this can also lookup items in the content cache
  * Handles redirect results
  * Runs the IContentFinders
* IContentFinder - these are replaceable today so in theory you could already replace the routing lookups for everything contained in this new interface - So then it's a decision of if it makes sense to have this new interface or just replace the existing IContentFinders. I guess it's easier to replace one service than many, but if we had persisted URLs maybe it wouldn't matter?
* IUrlProvider - outbound routing, generates URLs for content. I 'think' if we persisted the result of GetOtherUrls in a DB in a format that was easily queryable/indexed (possibly similar to the hashing mechanism done in the UrlRedirect table) it could work if executed on content save. 

If we had persisted URLs/routes then we could probably have an IContentFinder for that which would be the first one - this would satisfy the statement above that 

> IContentFinders won't really execute, only as a last resort if no persisted URLs are matched. 

But it would probably still be good to abstract out and split up the IPublishedRouter.TryRouteRequest